### ISSUE TITLE
CompileTraced can never validate for GGUF models: embedding lookup produces its output without an engine call, so instruction 0 reads a nil slot

### ISSUE THREAD

**Issue by dndungu** [zerfoo/zerfoo#994]
## Summary

Every generation logs this once, at the prefill→decode transition:

```
generate: CompileTraced plan validation failed, falling back to Compile:
instruction 0 (Mul): input tensors cannot be nil
```

It is **structural, not data-dependent**: the traced plan can never validate for any GGUF architecture. The traced/megakernel path is therefore dead code that costs a full extra forward pass plus a failed validation run on the first decode of every process, and emits an alarming log line on every run.

Split out of #990 during triage. It is **not** the cause of that bug — see the [analysis there](https://github.com/zerfoo/zerfoo/issues/990) — but it is real on its own.

## Root cause

`embeddingLookupNode.Forward` builds its output on the CPU path with plain Go — a hand-rolled row copy into a `[]T` and `tensor.New(...)` — making **zero engine calls** (`inference/arch_llama.go:322-341`).

`CompileTraced` builds its slot graph purely from ops recorded by the `EngineProxy` while it re-runs `Forward` on every node (`ztensor@v1.19.2/graph/compile.go:758-779`). Because the embedding produced its output without an engine call, no traced instruction *produces* that tensor. It only acquires a slot when the next traced op consumes it, via `tracer.SlotFor`. That slot is neither an input slot nor a frozen slot, so `PrepareSlots`/`RunInstructions` leave it nil and the run bails at `compile.go:545`:

```go
return nil, fmt.Errorf("instruction %d (%s): input tensors cannot be nil", i, inst.OpName)
```

`instruction 0 (Mul)` is simply the first engine op downstream of the embedding.

The embedding weight itself *is* registered (the node implements `EmbeddedFrozenProvider`), so this is not a missing-weight problem — it is the node's *output* that has no producer in the trace.

## Cost

- One full extra `Forward` over every node (the trace), plus a failed `ExecutionPlan.Run` validation, on the first decode step of every `Generator` (`generate/generator.go:301-320`, guarded by `planOnce`).
- The traced plan, and with it the megakernel emitter that depends on primitive-op decomposition, is unreachable for every GGUF model.
- The log line reads like a correctness failure. It was the first lead pulled in #990 and cost that investigation its opening day.

## Not a correctness problem

The `Compile` fallback is sound and is what actually runs. It reuses `g.memo` from the last real `Forward` instead of re-running it (`compile.go:580-602`), and its instructions call `n.Forward(ctx, ...)` with the *runtime* context, so the live KV cache is read from `genCtx` at `Run` time. Empirically, with #990's position bug fixed, generation through the fallback plan is token-identical to a control that re-prefills the whole sequence from a fresh cache at every step.

Worth noting as a latent hazard even so: `compileGraph` is invoked *after* `graph.Forward` returns the logits and *before* the caller samples them (`generate/session.go:455-458`, `generate/decode_step.go:46-58`). The trace re-runs every node's `Forward`. On CPU this does not clobber the pending logits — verified by logging the argmax before and after — but any future change that makes node outputs write into stable preallocated buffers would silently corrupt the token about to be sampled.

## Options

1. Teach the tracer about nodes that produce tensors outside the engine — e.g. let a node declare its output as a trace root, or record a synthetic `Gather` for the embedding lookup.
2. Detect such a node at compile time, skip `CompileTraced` entirely for that graph, and drop the message to debug level so it stops reading as a failure.

(1) unlocks the megakernel path for GGUF models; (2) is a few lines and removes the wasted work and the misleading log immediately. They are not exclusive.

## Reproduction

Any GGUF generation, CPU or GPU. Watch stderr:

```
GGUF=/var/lib/zerfoo/models/gemma3-1b-it-Q4_K_M.gguf  # any staged model
# inference.LoadFile(path, inference.WithDevice("cpu")); m.Generate(...)
```

**Comment by dndungu** [comment]
**Decision: remove (option 2). PR #996.** Traced compilation is now opt-in behind `ZERFOO_TRACED_PLAN=1`, off by default.

### The number that decided it: 341 dangling slots, not 1

The root cause in the writeup above is correct but understates the scope. `embeddingLookupNode.Forward` is not the only node that materialises a tensor in plain Go — it is just the first one an instruction reads. I instrumented the traced plan to count *dangling* slots: slots consumed by an instruction that are neither frozen, nor a plan input, nor an earlier instruction's output.

| model | dangling / total slots | frozen | inputs |
|---|---|---|---|
| gemma3-1b-it-Q4_K_M | **341 / 1701** | 288 | 1 |
| Qwen3-0.6B-Q8_0 | **283 / 1831** | 254 | 1 |

That is ~13 per transformer layer — RoPE tables, KV cache slices, attention masks, reciprocals. Option 1 (teach the tracer about the embedding) moves the failure from instruction 0 to instruction 1. It is not a node fix; it is a rewrite of how every architecture materialises intermediates.

### A working traced plan would be slower on CPU, not faster

The two plans are not the same work expressed differently:

```
traced   (1071 instr): Reshape=314 MatMul=235 Mul=157 Transpose=131 Add=52 Concat=26 Split=26 ...
compiled ( 185 instr): RMSNorm=53 FFN=26 ResidualRef=26 FusedNormAdd=26
                       GroupedQueryAttention=26 FusedAddRMSNorm=26 LMHead=1 EmbeddingLookup=1
```

Tracing decomposes the model into generic fp32 primitives and throws away every fused and quantised kernel the node-level plan uses. The megakernel emitter is the only genuine payoff, and it is CUDA codegen — nothing the CPU path can collect. So there was no performance story to justify the more interesting fix, and I am not going to manufacture one.

### Measured cost of the dead attempt (CPU, per generator)

| | gemma3-1b | Qwen3-0.6B |
|---|---|---|
| `CompileTraced` | 697–765 ms | 321 ms |
| failed validation `Run` | 37–50 µs | 34 µs |
| `Compile` fallback | 68–100 µs | 50 µs |

`Compile` is nearly free because it reuses `g.memo` and runs no forward pass. Everything above it is waste. Amortised over a long generation this is small — the honest headline is ~0.7 s once per generator plus a log line that reads like a correctness bug, not a throughput problem. That is exactly why removing it beats fixing it: the cost is too small to justify a rewrite, and too visible to leave alone.

### Scope

All 25 `SetEngineProxy` call sites in the repo are `inference/arch_*.go` builders. The old guard `proxy != nil` therefore selected precisely the set of graphs that cannot be traced — 100% of the time, CPU and GPU alike. There is no graph in this repository where the traced path validates, so nothing is lost by defaulting it off, and the flag restores the previous behaviour with no rebuild for the megakernel work.

### Verification

- **Log line gone.** Absent from every default run; present on every `ZERFOO_TRACED_PLAN=1` run.
- **Output unchanged.** Greedy decode, three prompts, token IDs compared (not just strings), flag on vs off in the same binary: gemma3-1b 33/31/26 tokens identical; Qwen3-0.6B 7/6/9 tokens identical.
- **Test red-proofed** against the current broken state, per L-0009/L-0018 and ztensor#182. With the new guard removed, `TestCompileGraph_TracedPlanOptIn/disabled_by_default` fails with the exact text from this issue (`instruction 0 (Mul): input tensors cannot be nil`). A second subcase asserts the opt-in still attempts tracing, so simply deleting the call would not pass either.

Recorded as **L-0022** in `docs/lore.md`, including the trigger "any proposal to just make the embedding emit a `Gather`", so this does not get re-litigated after the next context roll.

The latent hazard noted in the writeup — `compileGraph` running between `graph.Forward` and sampling, where a future move to stable preallocated node buffers would corrupt the pending token — is unaffected by this change and still stands. It is now cheaper to trip over (one `Compile`, no trace forward) but not eliminated; worth its own issue if anyone touches node output buffering.
